openmdao/core/problem.py

```diff
@@ -531,11 +567,170 @@ def run(self):
+
+        if rec_constraints:
+            myconstraints = all_constraints
+
+        # get the includes that were requested for this Driver recording
+        if incl:
+            # The my* variables are sets
+
+            # First gather all of the desired outputs
+            # The following might only be the local vars if MPI
+            mysystem_outputs = {n for n in model._outputs
+                                if check_path(n, incl, excl)}
+
+            # If MPI, and on rank 0, need to gather up all the variables
+            #    even those not local to rank 0
+            if MPI:
+                all_vars = model.comm.gather(mysystem_outputs, root=0)
+                if MPI.COMM_WORLD.rank == 0:
+                    mysystem_outputs = all_vars[-1]
+                    for d in all_vars[:-1]:
+                        mysystem_outputs.update(d)
+
+            # de-duplicate mysystem_outputs
+            mysystem_outputs = mysystem_outputs.difference(all_desvars, all_objectives,
+                                                           all_constraints)
+
+        # if rec_inputs:
+        #     root = self.model
+        #     myinputs = {n for n in root._inputs
+        #                 if check_path(n, incl, excl)}
+        #
+        #     if MPI:
+        #         all_vars = root.comm.gather(myinputs, root=0)
+        #         if MPI.COMM_WORLD.rank == 0:
+        #             myinputs = all_vars[-1]
+        #             for d in all_vars[:-1]:
+        #                 myinputs.update(d)
+
+        if MPI:  # filter based on who owns the variables
+            # TODO Eventually, we think we can get rid of this next check. But to be safe,
+            #       we are leaving it in there.
+            if not model.is_active():
+                raise RuntimeError("RecordingManager.startup should never be called when "
+                                   "running in parallel on an inactive System")
+            rrank = self.comm.rank
+            rowned = model._owning_rank
+            mydesvars = [n for n in mydesvars if rrank == rowned[n]]
+            myobjectives = [n for n in myobjectives if rrank == rowned[n]]
+            myconstraints = [n for n in myconstraints if rrank == rowned[n]]
+
+        self._filtered_vars_to_record = {
+            'des': mydesvars,
+            'obj': myobjectives,
+            'con': myconstraints,
+        }
+
+        self._rec_mgr.startup(self)
+        if self.recording_options['record_metadata']:
+            self._rec_mgr.record_metadata(self)
+
+    def add_recorder(self, recorder):
+        """
+        Add a recorder to the problem.
+
+        Parameters
+        ----------
+        recorder : BaseRecorder
+           A recorder instance.
+        """
+        self._rec_mgr.append(recorder)
+
     def cleanup(self):
         """
         Clean up resources prior to exit.
         """
         self.driver.cleanup()
+        self._rec_mgr.close()
```

Comment:
Heads up, you will need to make some changes once Steve's current PR is merged, where close-> cleanup